examples/structured_data/classification_with_tfdf.py

```diff
@@ -0,0 +1,726 @@
+        )
+
+        print(
+            "Target encoding: Computing frequencies for feature values with negative tatgets..."
+        )
+        # Filter the data where target label is negative.
+        negative_indices = tf.where(condition=tf.math.logical_not(target_values))
+        negative_feature_values = tf.gather_nd(
+            params=feature_values, indices=negative_indices
+        )
+        # Compute how many time each feature value occured with a negative target label.
+        negative_frequency = tf.math.unsorted_segment_sum(
+            data=tf.ones(
+                shape=(negative_feature_values.shape[0], 1), dtype=tf.dtypes.int32
+            ),
+            segment_ids=negative_feature_values,
+            num_segments=unique_feature_values.shape[0],
+        )
+
+        print("Target encoding: Storing target encoding statistics...")
+        self.positive_frequency_lookup = tf.constant(positive_frequency)
+        self.negative_frequency_lookup = tf.constant(negative_frequency)
+
+        self._is_adapted = True
+
+    def reset_state(self):
+        if self.built:
+            self.positive_frequency_lookup = None
+            self.negative_frequency_lookup = None
+
+    def call(self, inputs):
+        # data is expected to be a integer numpy array ot a Tensor shape [num_exmples, 1].
+        # This include the feature values for a given feature in the dataset.
+
+        # Raise an error of the target encoding statistics are not computed.
+        if (
+            self.positive_frequency_lookup == None
+            or self.negative_frequency_lookup == None
+        ):
+            raise ValueError(
+                f"You need to call the adapt method to compute target encoding statistics."
+            )
+
+        inputs = self._standardize_inputs(inputs)
+
+        # Fix and validate inputs shape.
+        if inputs.shape.rank == 1:
+            inputs = tf.expand_dims(inputs, axis=-1)
+        elif inputs.shape.rank > 2:
+            raise ValueError(
+                f"inputs is expected to be of rank 2, got {data.shape.rank}"
+            )
+        # Validate the dimensions of the second axis.
+        if inputs.shape[-1] != 1:
+            raise ValueError(
+                f"The second axis of inputs is expected to have 1 dimension, got {data.shape[-1]}"
+            )
+        # Lookup positive frequencies for the input feature values.
+        positive_fequency = tf.cast(
+            tf.gather_nd(self.positive_frequency_lookup, inputs),
+            dtype=tf.dtypes.float32,
+        )
+        # Lookup negative frequencies for the input feature values.
+        negative_fequency = tf.cast(
+            tf.gather_nd(self.negative_frequency_lookup, inputs),
+            dtype=tf.dtypes.float32,
+        )
+        # Compute positive probability for the input feature values.
+        positive_probability = positive_fequency / (
+            positive_fequency + negative_fequency
```

Comment:
This division will be unstable for rare categorical values. With target mean, it works well to add a correction term controlled by an hyperparameter.

```python
correction =  1
positive_probability = positive_frequency / (positive_frequency + negative_frequency + correction)
```